File: 준영/LV2/땅따먹기.py
def solution(land):
    answer = 0

    N = len(land)
    M = len(land[0])

    d_arr = [[0 for _ in range(M)] for _ in range(N)]
    
    for i in range(M):
        d_arr[0][i] = land[0][i]

    for i in range(1, N):
        for j in range(M):
            max_temp = 0
            for k in range(M):
                if j == k: continue

                if max_temp < d_arr[i-1][k]:
                   max_temp = d_arr[i-1][k]
            d_arr[i][j] = land[i][j] + max_temp
    answer = max(d_arr[i])
    return answer

# land = [[1,2,3,5],[5,6,7,8],[4,3,2,1]]
# solution(land)
land = [[1,1,1,1],[2,2,2,3],[3,3,3,6],[4,4,4,7]]
solution(land)

File: 준영/프로그래머스/LV2/땅따먹기.js
function solution(land) {
  var answer = 0
  
  const N = land.length

  const d_arr = Array.from(Array(N), () => Array(4).fill(0))
  
  for (let i = 0; i < 4; i++) {
    d_arr[0][i] = land[0][i]
  }

  for (let i = 1; i < N; i++) {
    for (let j = 0; j < 4; j++) {
      var max_temp = 0
      for (let k = 0; k < 4; k++) {
        if (j === k) continue

        if (max_temp < d_arr[i-1][k]) {
          max_temp = d_arr[i-1][k]
        }
        d_arr[i][j] = land[i][j] + max_temp
      }
    }
  }

  for (let i = 0; i < 4; i++) {
    if (d_arr[N-1][i] > answer) {
      answer = d_arr[N-1][i]
    }
  }
  
  return answer
}

land = [[1,1,1,1],[2,2,2,3],[3,3,3,6],[4,4,4,7]]
solution(land)